sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -140,25 +140,55 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
   require(list != null, "list should not be null")
   override def checkInputDataTypes(): TypeCheckResult = {
     list match {
-      case ListQuery(sub, _, _) :: Nil =>
+      case (l @ ListQuery(sub, children, _)) :: Nil =>
         val valExprs = value match {
           case cns: CreateNamedStruct => cns.valExprs
           case expr => Seq(expr)
         }
-        if (valExprs.length != sub.output.length) {
-          TypeCheckResult.TypeCheckFailure(
+
+        // SPARK-21759:
+        // It is possibly that the subquery plan has more output than value expressions, because
+        // the condition expressions in `ListQuery` might use part of subquery plan's output.
+        // For example, in the following query plan, the condition of `ListQuery` uses value#207
+        // from the subquery query. For now the size of output of subquery is 2, the size of value
+        // is 1.
+        //
+        //   Filter key#201 IN (list#200 [(value#207 = min(value)#204)])
+        //   :  +- Project [key#206, value#207]
+        //   :     +- Filter (value#207 > val_9)
```

Comment:
The example in PR description is constructed later. This example is I encountered in `subquery_in_having.q`. I'll make it consistent.